// plcm-encryption/Gruntfile.js
//   © , Inc. 2012-2015. All Rights Reserved

// Gruntfile.js
module.exports = function (grunt) {
    // load all grunt tasks matching the ['grunt-*', '@*/grunt-*'] patterns
    require('load-grunt-tasks')(grunt);

    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),
        shell: {
            options: {
                stdout: true
            },
            npm_install: {
                command: 'npm install'
            }
        },
        jshint: {
            all: ['*.js']
        },
        jasmine_nodejs: {
            options: {
                specNameSuffix: 'spec.js',
                stopOnFailure: false,
                // configure one or more built-in reporters
                reporters: {
                    console: {
                        colors: true
                    }
                }
			},
            all: {
                specs: ['lib/*']
            }
        },
        watch: {
            files: ['lib/*'],
            tasks: ['jasmine_nodejs']
        }
    });

    grunt.registerTask('test', ['jshint', 'jasmine_nodejs']);
    grunt.registerTask('default', ['test']);
};
